watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack_conc_recirc.py

```diff
@@ -0,0 +1,460 @@
+    m.fs.feed = Feed(property_package=m.fs.properties)
+    m.fs.sepa0 = Separator(
+        property_package=m.fs.properties,
+        outlet_list=["to_dil_in", "to_conc_in0"],
+    )
+    m.fs.mix0 = Mixer(
+        property_package=m.fs.properties,
+        energy_mixing_type=MixingType.none,
+        momentum_mixing_type=MomentumMixingType.none,
+        inlet_list=["from_feed", "from_conc_out"],
+    )
+
+    m.fs.pump0 = Pump(property_package=m.fs.properties)
+    m.fs.pump0.ratioP_calculation.deactivate()
+    m.fs.pump1 = Pump(property_package=m.fs.properties)
+    m.fs.pump0.ratioP_calculation.deactivate()
+
+    # Add electrodialysis (ED) stacks
+    m.fs.EDstack = Electrodialysis1D(
+        property_package=m.fs.properties,
+        operation_mode=ElectricalOperationMode.Constant_Voltage,
+        finite_elements=20,
+        has_pressure_change=True,
+        has_nonohmic_potential_membrane=True,
+        has_Nernst_diffusion_layer=True,
+        limiting_current_density_method=LimitingCurrentDensityMethod.Theoretical,
+        pressure_drop_method=PressureDropMethod.Darcy_Weisbach,
+        hydraulic_diameter_method=HydraulicDiameterMethod.spacer_specific_area_known,
+        friction_factor_method=FrictionFactorMethod.Gurreri,
+    )
+    m.fs.sepa1 = Separator(
+        property_package=m.fs.properties,
+        outlet_list=["to_disp", "to_conc_in1"],
+    )
+    m.fs.prod = Product(property_package=m.fs.properties)
+    m.fs.disp = Product(property_package=m.fs.properties)
+
+    # touch necessary variables to ensure they are constructed and solved
+    m.fs.prod.properties[0].flow_vol_phase[...]
+    m.fs.disp.properties[0].flow_vol_phase[...]
+
+    # costing
+    m.fs.EDstack.costing = UnitModelCostingBlock(flowsheet_costing_block=m.fs.costing)
+    m.fs.pump0.costing = UnitModelCostingBlock(flowsheet_costing_block=m.fs.costing)
+    m.fs.pump1.costing = UnitModelCostingBlock(flowsheet_costing_block=m.fs.costing)
+    m.fs.costing.cost_process()
+    m.fs.costing.add_annual_water_production(
+        m.fs.prod.properties[0].flow_vol_phase["Liq"]
+    )
+    m.fs.costing.add_LCOW(m.fs.prod.properties[0].flow_vol)
+    m.fs.costing.add_specific_energy_consumption(
+        m.fs.prod.properties[0].flow_vol_phase["Liq"]
+    )
+
+    # add extra variables and constraints
+    m.fs.recovery_vol_H2O = Var(
+        initialize=0.5,
+        bounds=(0, 1),
+        domain=NonNegativeReals,
+        units=pyunits.dimensionless,
+        doc="flowsheet level water recovery calculated by volumeric flow rate",
+    )
+    m.fs.mem_area = Var(
+        initialize=1,
+        bounds=(0, 1e3),
+        units=pyunits.meter**2,
+        doc="Total membrane area for cem (or aem) in one stack",
+    )
+    m.fs.prod_salinity = Var(
+        initialize=1, bounds=(0, 1000), units=pyunits.kg * pyunits.meter**-3
+    )
+    m.fs.disp_salinity = Var(
+        initialize=1, bounds=(0, 1e6), units=pyunits.kg * pyunits.meter**-3
+    )
```

Comment:
Also, why do you need these variables? I might be mistaken, but can't you just touch conc_mass_phase_comp on the corresponding Product blocks?

Reply:
It was like you suggested before.  But I found when I add up the concentration to express a molecular salinity, doing it in this way direct the solver to minimize the error of the sum of two species.  In others words, minimizing two individual errors may not minimize the sum of the two if that makes sense.